src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -482,9 +485,139 @@ EventEngine::TaskHandle PosixEventEngine::RunAfterInternal(
+    absl::string_view default_port, Duration timeout) {
+  GrpcAresHostnameRequest* request = new GrpcAresHostnameRequest(
+      name, default_port, timeout,
+      absl::bind_front(&PosixEventEngine::PosixDNSResolver::CreateEventHandle,
+                       this),
+      event_engine_);
+  const absl::Status status =
+      request->Initialize(options_.dns_server, /*check_port=*/true);
+  if (!status.ok()) {
+    // Report back initialization failure through on_resolve.
+    on_resolve(status);
+    return LookupTaskHandle{};
+  }
+  LookupTaskHandle handle{reinterpret_cast<intptr_t>(request),
+                          aba_token_.fetch_add(1)};
+  {
+    absl::MutexLock lock(&mu_);
+    inflight_requests_.insert(handle);
+  }
+  request->Start([on_resolve = std::move(on_resolve), handle, this](
+                     absl::StatusOr<std::vector<EventEngine::ResolvedAddress>>
+                         result) mutable {
+    {
+      absl::MutexLock lock(&mu_);
+      // on_resolved called, no longer inflight.
+      GPR_ASSERT(inflight_requests_.erase(handle) == 1);
+    }
+    on_resolve(std::move(result));
+  });
+  return handle;
+}
+
+LookupTaskHandle PosixEventEngine::PosixDNSResolver::LookupSRV(
+    LookupSRVCallback on_resolve, absl::string_view name, Duration timeout) {
+  GrpcAresSRVRequest* request = new GrpcAresSRVRequest(
+      name, timeout,
+      absl::bind_front(&PosixEventEngine::PosixDNSResolver::CreateEventHandle,
+                       this),
+      event_engine_);
+  const absl::Status status =
+      request->Initialize(options_.dns_server, /*check_port=*/false);
+  if (!status.ok()) {
+    // Report back initialization failure through on_resolve.
+    on_resolve(status);
+    return LookupTaskHandle{};
+  }
+  LookupTaskHandle handle{reinterpret_cast<intptr_t>(request),
+                          aba_token_.fetch_add(1)};
+  {
+    absl::MutexLock lock(&mu_);
+    inflight_requests_.insert(handle);
+  }
+  request->Start(
+      [on_resolve = std::move(on_resolve), handle,
+       this](absl::StatusOr<std::vector<EventEngine::DNSResolver::SRVRecord>>
+                 result) mutable {
+        {
+          absl::MutexLock lock(&mu_);
+          // on_resolved called, no longer inflight.
+          GPR_ASSERT(inflight_requests_.erase(handle) == 1);
+        }
+        on_resolve(std::move(result));
+      });
+  return handle;
+}
+
+LookupTaskHandle PosixEventEngine::PosixDNSResolver::LookupTXT(
+    LookupTXTCallback on_resolve, absl::string_view name, Duration timeout) {
+  GrpcAresTXTRequest* request = new GrpcAresTXTRequest(
+      name, timeout,
+      absl::bind_front(&PosixEventEngine::PosixDNSResolver::CreateEventHandle,
+                       this),
+      event_engine_);
+  const absl::Status status =
+      request->Initialize(options_.dns_server, /*check_port=*/false);
+  if (!status.ok()) {
+    // Report back initialization failure through on_resolve.
+    on_resolve(status);
```

Comment:
This needs to be run asynchronously. Otherwise, we can expect a deadlock here in the client channel resolver attempts to resolve balancer hostnames.

```
event_engine_->Run([cb = std::move(on_resolved), status]() mutable {
  cb(status);
});
```